src/utils/async.ts

```diff
@@ -48,3 +50,43 @@ export async function asyncIterableToArray<T>(
   const iter = source[Symbol.asyncIterator]();
   return asyncIteratorToArray(iter);
 }
+
+export function mapEitherAsyncIterator<E, T, V>(
+  iter: AsyncIterator<Either<E, T>, Either<E, T>>,
+  f: (t: T) => V
+): AsyncIterator<V> {
+  return {
+    next: async () => {
+      while (true) {
+        const { done, value } = await iter.next();
+        if (done) {
+          return { done, value: undefined };
+        }
+        if (isRight(value)) {
+          return { done, value: f(value.value) };
+        }
+      }
+    }
+  };
+}
+
+export function flattenAsyncIterator<T>(
+  iter: AsyncIterator<ReadonlyArray<T>>
+): AsyncIterator<T> {
+  // tslint:disable-next-line: no-let
+  let index = 0;
+  // tslint:disable-next-line: no-let
+  let flattenArray: ReadonlyArray<T> = [];
+  return {
+    next: async () => {
+      if (flattenArray.length === index) {
+        const { done, value } = await iter.next();
+        if (done) {
+          return { done, value: undefined };
```

Comment:
the caller should probably be aware that the last returned value gets ignored

Reply:
How do you suggest to warning the caller? Could be enough with a comment description over the function?